client/layout/guided-tours/tours/jetpack-plugin-updates-tour/index.tsx

```diff
@@ -34,9 +35,43 @@ export const JetpackPluginUpdatesTour = makeTour(
 	>
 		<Step
 			name="init"
-			target=".plugin-item-jetpack .form-toggle__switch"
+			target={ JETPACK_TOGGLE_SELECTOR }
 			arrow="top-left"
 			placement="below"
+			wait={ () =>
+				new Promise( resolve => {
+					if ( document.querySelector( JETPACK_TOGGLE_SELECTOR ) ) {
+						return resolve();
+					}
+
+					const primaryContainer = document.querySelector( '#primary' );
+					if ( typeof MutationObserver === 'undefined' || ! primaryContainer ) {
+						return setTimeout( resolve, 2000 );
+					}
+
+					new MutationObserver( ( mutationRecords, observer ) => {
+						mutationRecords.some( mutationRecord => {
+							if ( mutationRecord.type === 'childList' && mutationRecord.addedNodes.length ) {
+								if (
+									Array.from( mutationRecord.addedNodes ).some(
+										node =>
+											node.nodeType === Node.ELEMENT_NODE &&
+											( node as Element ).classList.contains( 'plugin-item-jetpack' )
+									)
+								) {
+									resolve();
+									observer.disconnect();
+									return true;
+								}
+							}
+							return false;
+						} );
+					} ).observe( primaryContainer, {
+						childList: true,
+						subtree: true,
+					} );
```

Comment:
this is a nice use of `MutationObserver` but would it be conceptually easier if we just waited in a loop for the item to appear? also why do we immediately succeed if we find `JETPACK_TOGGLE_SELECTOR` but then wait for `plugin-item-jetpack` to appear?

```ts
wait={ () =>
	new Promise( resolve => {
		if ( document.querySelector( JETPACK_TOGGLE_SELECTOR ) ) {
			return resolve();
		}

		const waitForItem = () => {
			if ( document.querySelector( '#primary .plugin-item-jetpack' ) ) {
				return resolve();
			}

			requestIdleCallback( waitForItem );
		}
	} )
}
```

Reply:
While I always like a simple solution :grimacing: and agree that it would be specifically nice to express it by only refererring to `JETPACK_TOGGLE_SELECTOR`, this doesn't seem to work OOTB for me :sweat_smile: (I also tried moving the `requestIdleCallback` out of `waitForItem`.)

Reply:
👍 Good observation, I naively based this on what `step` does internally and didn't consider alternatives.

In 02cdaa347bab8f5068dfe889ec352de0a1e2ae8e I've implemented the same functionality with a simple `setTimeout`.

Reply:
oops - I obviously forgot to _call_ `waitForItem()` in my example.

the `setTimeout` is fine - did `requestIdleCallback()` not work?

Reply:
I didn't try `requestIdleCallback` for a couple of reasons:
- Browser support looks pretty good but not perfect. I'm not sure what the polyfill situation is.
- We want to show the tour step as early as possible, I want this to be eager.

Reply:
@sgomes didn't I see you add something with `requestIdleCallback` lately? can you help us understand our support situation with it?

@sirreal I would imagine that `requestIdleCallback` would fire significantly more frequently and earlier than 125ms…were you thinking the opposite?

Reply:
Happy to help! `requestIdleCallback` and `setTimeout` don't serve as exact replacements for each other; the semantics are different.

`setTimeout` is a fixed thing. It executes after at least `X` milliseconds, and at least in the following event loop (which is why `setTimeout(foo, 0)` executes on the next event loop and not synchronously).

`requestIdleCallback` is different; it'll call back when the browser is next idle and expects to be able to run your code without negatively impacting performance. Any tasks that you execute in an idle callback should try to execute in 50ms (which is not a problem here), or should otherwise chunk their work across multiple idle callbacks, by voluntarily releasing the main thread and scheduling another idle callback.

As such, you can't expect `requestIdleCallback` to fire on any particular interval; it all depends on what else the browser is doing, and what heuristics it uses to determine when it's a good time to fire the callbacks. You can, however, specify a timeout value on `requestIdleCallback` to try to get the browser to fire the callback earlier, even if that runs the risk of having an impact on performance.

In this situation, both `requestIdleCallback` and `setTimeout` seem like good approaches, at least taking into account the code here. The only thing that's happening on callback is a DOM read, which is fast and doesn't cause any negative side-effects, like layout invalidation. What happens after the `Promise.resolve()` is less clear, but that only happens once and it's behind a microtask anyway, IIRC. `requestIdleCallback` would be a bit better here, in my opinion, but not by much; you'd still probably want to add a "magic number" timeout.

As for support, it's pretty good. There would be no need for a polyfill; in this situation, falling back to a `setTimeout` is perfectly fine. For example (loosely tested):

```js
const hasRIC = typeof window !== 'undefined' && !! window.requestIdleCallback;

function waitAndCheck( fn, timeout ) {
  if ( hasRIC ) {
    const id = requestIdleCallback( fn, { timeout } );
    return () => cancelIdleCallback( id );
  }
  
  const id = setTimeout( fn, timeout );
  return () => clearTimeout( id );
}
```

This adds the callback to the right queue and returns a cancelling function — do remember to cancel the timeout when the component gets destroyed. The semantics for the timeout are different, as previously explained, in that one is "at most, if possible", and the other's "at least", but for this use-case the simple code above should be enough.

Let me know if you have any questions!

Reply:
Thanks both, this is helpful and informative.

I'm going to refrain from any adjustments here. The solution for this particular `Step` will likely be replaced by a framework solution in #33369, I've suggested following your advice there: https://github.com/Automattic/wp-calypso/pull/33369/files#r288491844